### ISSUE TITLE
Submission API

### ISSUE THREAD

**Issue by dotcs** [nils-braun/learningpage#7]
I think we should re-work the submission endpoints to give the user a better overview of her submissions.

I'd prefer to have those endpoints:
- `GET /content/<slug>/submissions`  
    This endpoint should return a list of all submissions that have been done in the past. All items should been returned whether they have been graded or not.
- `POST  /content/<slug>/submissions`
    This endpoint should create a new submission.

The list of submissions could look like this:
```
{
  maxScore: float,
  submissions: ISubmission[]
}
```

with ISubmission
```
{ date: string, graded: boolean, notebooks: INotebook, score: float }
```

and INotebook
```
{ name: string, score: float, feedbackUrl: string }
```

`score` in `ISubmission` should be the averaged(?) scored of all notebook scores. `maxScore` should be the maximum score of all submissions.

After `POST`ing a new submission the return value should also be the an `ISubmission` with `graded: false` since it will happen asynchronously.

Having such an API would allow to show a table in the front-end that contains information about all submissions which makes it easier for users to understand where the `maxScore` comes from and how they have been improved over time.

@nils-braun What do you think?

**Comment by nils-braun** [comment]
Thanks for bringing this up, @dotcs 

(just a small, more or less unrelated comment: the score of a submission is the sum of the scores of each notebook)

However, there is a reason why the current API only shows the current available submission.
The reason is, that the nbgrader database does not store more than that.

So, two possibilities:
* we keep it with only a single submission
* we need to store the submissions separately from the nbgrader database. This causes some synchronization hassle (because grading can be both automatic or manual in nbgrader). It is not impossible though. However before tackling this, I would like to clearly understand why it is needed: what does a user gain by seeing the full history? why should the next submission be worse than the one before (the feedback from nbgrader is very detailed)? why is a user interested in seeing the old scores while his new submission is graded? 
I understand that it is "nice to have", but if we look into other systems such as Coursera (the quiz feedback is lost once you start a new quiz) or Udacity (the capstone project feedback is lost, once you do a new submission) they have also not implemented it. 

I have even currently not implemented (on purpose) the fact, that a new submission only overrides the old submission of the score is better. I want people to understand what is correct and what not and not to submit until everything is 100%...

But that was clearly a design decision, so lets discuss!

**Comment by dotcs** [comment]
Thanks for the detailed answer. Okay, I see your point as well. We don't need to make things more complicated than they are and you're right. Coursera for example also only shows a single score - in their case the best one out of all submissions. We can do the same.

For the student I guess it's still interesting have a historic overview over all submissions and the grader output which will help to better understand the progress. Also from a analytical perspective for the tutor it could be interesting to see about how the average student did, how many submissions were necessary, which parts of the assignments were hard so solve, which one were easy for the students, etc. Keeping all submissions and their results would open the door for this.

To me it seems that we are still missing some parts. The way it is implemented right now does not allow to show the user how many submissions have been done in total. It also does not allow to implement a maximum number of attempts per day. Or do I miss something here?

**Comment by nils-braun** [comment]
Yes, that is correct.
I will need to think about a way how to implement the submission separate from nbgrader then. Should be possible.

Thanks for the suggestions.

**Comment by nils-braun** [comment]
Is the proposal implemented in #9 good to solve this issue?

**Comment by dotcs** [comment]
Absolutely. I think the endpoint `/content/<content_slug>/submissions` does what we need. I would consider this issue to be resolved once the PR has been merged.

**Comment by nils-braun** [comment]
Solved via #9
